### ISSUE TITLE
AArch64: Implement effective address of DC ZVA cache zero operation

### ISSUE THREAD

**Issue by mwvantol** [DynamoRIO/dynamorio#4329]
While running the drcachesim component, AArch64 data cache operations such as DC ZVA are not marked as making a memory access. However, these are often used in userspace in for example bzero() implementations to efficiently zero out memory.

**Comment by derekbruening** [comment]
Presumably this is an AArch64 decoder bug where `DC ZVA` has the wrong operand type and so is seen by tools like drcachesim's tracer as not accessing memory.  @AssadHashmi hoping you can take a look?

**Comment by mwvantol** [comment]
There are also several other cache operations that can execute in EL0/userspace. These would be to invalidate or clean (write back) data from the caches;
IC IVAU

DC CVAC
DC CVAU
DC CVAP
DC CIVAC

Note that all these instructions are variants of the SYS instruction, depending on the immediate value used (and there are many more that are privileged). More valid variations exist when certain architecture extensions are present, such as v8.5 memory tagging, which adds variants of these to deal with the data and tags.

I only found a mention of IC IVAU in the DynamoRIO source code as it needs to deal with that for self-modifying code. I haven't seen any of the above (except DC ZVA) in the workloads I'm looking at so far, so I don't know how DynamoRIO behaves around them. DC ZVA is certainly the most commonly used.

**Comment by AssadHashmi** [comment]
Hi @mwvantol, do you have a small test case which replicates the DC ZVA omission?

**Comment by mwvantol** [comment]
I don't have anything at hand, but will try to create one, and also include the other cacheops I listed above.

**Comment by mwvantol** [comment]
I've written and ran the following small program wrapped with the C runtime; you'll also find DC ZVA's in the C runtime initialization code. I confirmed that for none of the IC/DC instructions I see the effective memory address on the drcachesim tracing interface.
```
    .arch armv8.1-a
    .file   "icdc-test.c"
    
    .text
    .align 2
    .global main

main:
    adr x1, flushaddr
    dc zva, x1
    ic ivau, x1
    dc cvac, x1
    dc cvau, x1
    dc civac, x1

    ret


    .data
    .align 10
    .type   flushaddr, %object
    .size   flushaddr, 8

flushaddr:
    .xword 0x1

```
Please find the binary attached, or compile the above yourself by simply passing it as a .s file to GCC.
[icdc_test.zip](https://github.com/DynamoRIO/dynamorio/files/4819115/icdc_test.zip)

**Comment by derekbruening** [comment]
Others have recently hit this when simulating AArch64-gathered traces as well, including with just the sample trace at https://github.com/DynamoRIO/drmemtrace_samples

**Comment by abhinav92003** [comment]
Hi. Just wanted to check if there's any progress on this. I hit this issue while handling AArch64 flush instructions on #4328.

**Comment by AssadHashmi** [comment]
> Hi. Just wanted to check if there's any progress on this. I hit this issue while handling AArch64 flush instructions on #4328.

Sorry, no fix yet. I'll make time to look at this during the week. Is it blocking #4328?

**Comment by derekbruening** [comment]
This is affecting every new project that starts using AArch64 traces for simulation, so it would be good to get it fixed.

Are there other missing memory references in any of the undecoded "xx" opcodes?

**Comment by AssadHashmi** [comment]
> This is affecting every new project that starts using AArch64 traces for simulation, so it would be good to get it fixed.

I'll do a PR tomorrow.

> Are there other missing memory references in any of the undecoded "xx" opcodes?

I've not looked at the "xx" bucket for a while. I'll check.

**Comment by nuviaram** [comment]
We are attempting to leverage Dynamo Rio for instruction tracing for chip-level / cache simulations. The lack of addresses capture for DCZVA is a blocker for us. Fixing this bug will be much appreciated.

**Comment by derekbruening** [comment]
If this issue covers just marking DC ZVA as a store but not making it aligned (which is #4400) then I think we can close this as PR #5235 implemented that, and leave #4400 open.
